crates/typst-syntax/src/parser.rs

```diff
@@ -1494,23 +1497,70 @@ fn pattern_leaf<'s>(
     }
 }
 
-/// Manages parsing of a stream of tokens.
+/// A single parsed Token.
+#[derive(Clone)]
+struct Token<'s> {
+    kind: SyntaxKind,
+    text: &'s str,
+}
+
+/// The starting point of any trivia (whitespace/comments) preceding the current token.
+#[derive(Clone)]
+struct TriviaStart {
+    /// Number of preceding trivia `nodes` (for avoiding trivia when wrapping).
+    num: usize,
+    /// Offset into `text` of the trivia start (for moving the lexer's cursor back).
+    offset: usize,
+}
+
+/// Manages parsing of a stream of tokens into a tree of SyntaxNodes.
+///
+/// This acts like a bottom-up (shift-reduce) parser, where we investigate a current token
+/// and either:
+/// 1. Eat (shift) another token, putting `current` into the `nodes` vector.
+/// 2. Wrap (reduce) the nodes up to a marker (excluding `current`) into a corresponding
+///    SyntaxKind.
```

Comment:
I'm not sure this is accurate. As far as I known, recursive descent parser is top-down and very different from shift-reduce/bottom-up. How we manage the node collection is sort of conceptually separate from the parsing structure.

Reply:
Agreed, was conflating the way we parse with the way we build the nodes when writing this. Trying to note the similarity to shift-reduce parsers isn't really helpful here.

How about this rewrite?
```
/// Manages parsing of a stream of tokens into a tree of SyntaxNodes.
///
/// The implementation presents an interface that can investigate a current token and take
/// one of the following actions:
/// 1. Eat a token, placing `current` into the `nodes` vector as a `LeafNode` and
///    preparing a new `current` by calling into the lexer.
/// 2. Wrap nodes from a marker to the end of `nodes` (excluding `current`) into an
///    `InnerNode` of a specific SyntaxKind.
/// 3. Produce or convert nodes into an `ErrorNode` when something expected is missing or
///    something unexpected is found.
///
/// Overall the Parser produces a nested tree of SyntaxNodes as a *Concrete* Syntax Tree.
/// The raw Concrete Syntax Tree should contain the entire source text, and is used as-is
/// for e.g. syntax highlighting and IDE features. In `ast.rs` the CST is interpreted as a
/// lazy view over an *Abstract* Syntax Tree. The AST module skips over irrelevant tokens
/// -- whitespace, comments, code parens, commas in function args, etc. -- as it iterates
/// through the tree.
```